src/framework.js

```diff
@@ -0,0 +1,194 @@
+
+    componentFactory: function( nodeDef, wrapper ) { 
+        
+        var compFactory = require('./noflo-component-factory');
+        return compFactory(
+                _.defaults( {
+                     // Add an ondata function to be invoked every time we receive new data
+                     inPorts: _.mapObject(nodeDef.inPorts, function( port, portName ) {  
+                         return _.extend({
+                             ondata: function(payload, socketIndex) {
+                                   var that = this; 
+                                   return ondata.call( that, payload, portName, socketIndex, port );
+                             }
+                         }, port);
+                     })
+                 }, nodeDef ),
+                { rpf: {      // extend default noflo component node instance with framework rpf object 
+                     vnis: vnis,
+                     wrapper: wrapper
+                  }
+                } 
+        );
+    }
+
+}; // module.exports
+
+
+var ondata = function( payload, portName, socketIndex, port ) {
+
+     // socketIndex is only defined for addressable/multi ports so set default of 0 if not defined.  
+     // If there is only one input, we can just default to socket index 0, so if not defined, 
+     // do this for now.   We need to discuss whether addressable should just be required on our 
+     // pipeline ports.  IMHO, I think it probably should be to minimize potential user-error in this area
+     // and simplify the interface our users work with.
+     socketIndex = socketIndex || 0;
+
+     var vnid = payload.vnid || '';
+     var vni = vnis.get.call( this.nodeInstance, vnid );
+     vni.inputStates.set.call( vni.node, 
+                               portInfo.call( vni.node, portName, socketIndex ),
+                               stateFactory( vnid, payload ) );
+
+     // check if should run updater & run if necessary
+     if ( shouldRunUpdater( vni ) ) { 
+        var data = this.nodeInstance.rpf.wrapper.fRunUpdater( vni.inputStates );
+        if ( data ) {
+            var outputState = stateFactory( vnid, data );
+            vni.outputState = outputState;
+
+            // TODO: Discuss whether this should be in the access layer instead of here
+            this.nodeInstance.outPorts.output.send( outputState )
+            this.nodeInstance.outPorts.output.disconnect();
+
+            // Clean up input states
+            vni.inputStates.deleteAll();
+        }
+     }
+
+     return;
+};
+
+/** 
+ * Check the input to see if we have all data required to run the updater or not. 
+ * This is simply a default updater policy - we may have other policies in the future.
+ * 
+ * @param vni virtual node instance whose input states are to be checked.
+ *
+ * @return true if we have received data on all required input port edges
+ */
+function shouldRunUpdater( vni ) { 
+
+    var requiredInPorts = requiredInputPorts.call( vni.node );
```

Comment:
Actually it needs to check all *attached* ports -- not required ports.   

Also, one other issue we'll have to deal with: If a port is not required and is not multi and has nothing attached, there will still be a corresponding updater parameter, which should receive an undefined value.  The easiest way to deal with this case might be to create a fake (default) IIP to pass in, with undefined data.